x-pack/plugin/otel-data/src/main/java/org/elasticsearch/xpack/oteldata/otlp/OTLPMetricsTransportAction.java

```diff
@@ -62,7 +80,131 @@ public OTLPMetricsTransportAction(
 
     @Override
     protected void doExecute(Task task, MetricsRequest request, ActionListener<MetricsResponse> listener) {
-        listener.onResponse(new MetricsResponse(RestStatus.NOT_IMPLEMENTED, new BytesArray(new byte[0])));
+        BufferedByteStringAccessor byteStringAccessor = new BufferedByteStringAccessor();
+        DataPointGroupingContext context = new DataPointGroupingContext(byteStringAccessor);
+        try {
+            var metricsServiceRequest = ExportMetricsServiceRequest.parseFrom(request.exportMetricsServiceRequest.streamInput());
+            context.groupDataPoints(metricsServiceRequest);
+            if (context.totalDataPoints() == 0) {
+                handleEmptyRequest(listener);
+                return;
+            }
+            BulkRequestBuilder bulkRequestBuilder = client.prepareBulk();
+            MetricDocumentBuilder metricDocumentBuilder = new MetricDocumentBuilder(byteStringAccessor);
+            context.consume(dataPointGroup -> addIndexRequest(bulkRequestBuilder, metricDocumentBuilder, dataPointGroup));
+            if (bulkRequestBuilder.numberOfActions() == 0) {
+                handleEmptyBulk(listener, context);
+                return;
+            }
+
+            bulkRequestBuilder.execute(new ActionListener<>() {
```

Comment:
I'm planning some unit tests that mock/spy `client` to test the different failure scenarios.